lib/ip.js

```diff
@@ -9,38 +9,51 @@ ip.toBuffer = function toBuffer(ip, buff, offset) {
 
   var result;
 
-  if (/^::ffff:(\d{1,3}\.){3,3}\d{1,3}$/.test(ip)) {
-    ip = ip.replace(/^::ffff:/, '');
-  }
-
-  if (/^(\d{1,3}\.){3,3}\d{1,3}$/.test(ip)) {
+  if (this.isV4Format(ip)) {
     result = buff || new Buffer(offset + 4);
     ip.split(/\./g).map(function(byte) {
       result[offset++] = parseInt(byte, 10) & 0xff;
     });
-  } else if (/^[a-f0-9:]+$/.test(ip)) {
-    var s = ip.split(/::/g, 2);
-    var head = (s[0] || '').split(/:/g, 8);
-    var tail = (s[1] || '').split(/:/g, 8);
-
-    if (tail.length === 0) {
-      // xxxx::
-      while (head.length < 8) head.push('0000');
-    } else if (head.length === 0) {
-      // ::xxxx
-      while (tail.length < 8) tail.unshift('0000');
-    } else {
-      // xxxx::xxxx
-      while (head.length + tail.length < 8) head.push('0000');
+  } else if (this.isV6Format(ip)) {
+    var sections = ip.split(':', 8);
+
+    for (var i = 0; i < sections.length; i++) {
+      var isv4 = this.isV4Format(sections[i]);
+      var v4Buffer;
+
+      if (isv4) {
+        v4Buffer = this.toBuffer(sections[i]);
+        sections[i] = v4Buffer.slice(0, 2).toString('hex');
+      }
+
+      if (v4Buffer && ++i < 8) {
+        sections.splice(i, 0, v4Buffer.slice(2, 4).toString('hex'));
+      }
+    }
+
+    if (sections[0] === '') {
+      while (sections.length < 8) sections.unshift('0');
+    } else if (sections[sections.length] === '') {
```

Comment:
Wow, can't believe this skipped right over me. Yeah, should be length-1. Will change that as soon as I get the chance.

The difference between them is one is to add '0' entries to the left side, another to the right, and the third is to add them to the middle. Since the string is split by `':'` tokens, an empty string indicates where those entries go.